nmpolicy/internal/resolver/visitor.go

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2021 NMPolicy Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *	  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package resolver
+
+type stateVisitor interface {
+	visitLastMap(path, map[string]interface{}) (interface{}, error)
+	visitLastSlice(path, []interface{}) (interface{}, error)
+	visitMap(path, map[string]interface{}) (interface{}, error)
+	visitSlice(path, []interface{}) (interface{}, error)
+}
+
+func visitState(p path, inputState interface{}, v stateVisitor) (interface{}, error) {
+	originalMap, isMap := inputState.(map[string]interface{})
+	if isMap {
+		if p.hasMoreSteps() {
+			if p.currentStep.Identity == nil {
+				return nil, pathError(p.currentStep, "unexpected non identity step for map state '%+v'", originalMap)
+			}
+			return v.visitMap(p, originalMap)
+		}
+		return v.visitLastMap(p, originalMap)
+	}
+
+	originalSlice, isSlice := inputState.([]interface{})
+	if isSlice {
+		if p.hasMoreSteps() || p.currentStep.Number == nil {
```

Comment:
It affects filter and replace, normally an expression would be explicit about recursive location so it would be something like 
```
routes.running.*.destination
```

So the step that is on the slice is not the last and contains a clear instruction about what to do `*` this means iterate them all, but instead we have
```
routes.running.destination
```

So although is the last step it has to delegate on normal visitSlice in case the step is a identity, have just put that code at both replace and filter.